### PR TITLE
BFD-2997: Introduce Environment Multi Region Keys

### PR DESCRIPTION
**Pull request by malessi**
**JIRA Ticket:**
[BFD-2997](https://jira.cms.gov/browse/BFD-2997)

**User Story or Bug Summary:**

As a BFD Engineer, I want a multi region key (MRK) to be used when encrypting sensitive values so that I can continue to access these values in the event of a disaster.

---

### What Does This PR Do?

This PR:

- Adds a new `mgmt`-level "config" Multi-Region Key (MRK) with a corresponding replica key in `us-west-2` called `bfd-mgmt-config-cmk` to `mgmt`'s `base_config` module
- Updates all `mgmt` `base_config` sensitive SSM parameters to use the aforementioned MRK for encryption instead of `bfd-mgmt-cmk`
- Updates all IAM policies that had permissions to `bfd-mgmt-cmk` to either:
  - Have _just_ permissions to `bfd-mgmt-config-cmk`, if it was obvious that the only usage of the `bfd-mgmt-cmk` was to decrypt sensitive SSM parameters
  - Have permission to use _both_ `bfd-mgmt-config-cmk` and `bfd-mgmt-cmk`, if it was not obvious whether or not the policy required both
- Adds a new per-environment "config" MRK, with corresponding replica in `us-west-2`, called `bfd-${env}-config-cmk` to the `base` Terraservice
- Updates all `base` sensitive SSM parameters to use the aforementioned, per-environment MRK for encryption
- Updates all IAM policies that had permissions to `bfd-${env}-cmk` to either:
  - Have _just_ permissions to `bfd-${env}-config-cmk`, if it was obvious that the only usage of the `bfd-${env}-cmk` was to decrypt sensitive SSM parameters
  - Have permission to use _both_ `bfd-${env}-config-cmk` and `bfd-${env}-cmk`, if it was not obvious whether or not the policy required both
- Updates `cipher` to support MRKs via the MRK-aware key provider
  - Note that [the MRK-aware key provider does not differ in behavior when compared to the default key provider if a Single-Region Key is provided](https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/configure.html#config-mrks):
    > For single-Region keys, the multi-Region-aware symbol behaves just like the single-Region AWS KMS keyring and master key provider. It attempts to decrypt ciphertext only with the single-Region key that encrypted the data

This PR has been validated by:

_Any changes to our infrastructure made as part of validating this PR have been reverted_

- Targeted `terraform apply`ing changes to `mgmt`'s `base_config` to create _just_ the MRKs, verifying that the MRKs were created as expected
- `terraform apply`ing `base` to `test`, verifying that all changes apply as expected
- `terraform apply`ing `base` to an ephemeral environment, `2997-test`, verifying that all changes apply as expected and sensitive SSM parameters are encrypted using the seed environment's MRK
- `terraform plan`ning `server` in `test`, verifying that the plan has no unexpected changes
- `terraform plan`ning `mgmt`, verifying that the plan has no unexpected changes
- Using `cipher` to encrypt using a SRK and then decrypt the same file using the same SRK, verifying that the file is decrypted successfully
- Using `cipher` to encrypt using a MRK and then decrypt the same file using the same MRK, verifying that the file is decrypted successfully
- Using `cipher` to encrypt using a MRK and then decrypt the same file using the MRK's `us-west-2` replica key, verifying that the file is decrypted successfully
- Using the AWS CLI to retrieve a sensitive parameter from SSM encrypted with one of the new MRKs, verifying that I was able to successfully decrypt it and view its value

**DUE TO THE NATURE OF THESE CHANGES, I HAVE OPTED NOT TO _FULLY_ APPLY CHANGES TO ANY TERRASERVICES OUTSIDE OF `base`. IT IS _VERY_ LIKELY THAT THE CHANGES TO IAM POLICIES WILL CAUSE FAILURES TO APPLY THAT WILL REQUIRE MORE CAREFUL DEPLOYMENT**

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

- Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

- If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

  - Does this PR add any new software dependencies?
    - [ ] Yes
    - [x] No
  - Does this PR modify or invalidate any of our security controls?
    - [ ] Yes
    - [x] No
  - Does this PR store or transmit data that was not stored or transmitted before?
    - [ ] Yes
    - [x] No

- If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
  - Do you think this PR requires additional review of its security implications for other reasons?
    - [ ] Yes
    - [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

- N/A

### Submitter Checklist

I have gone through and verified that...:

- [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
- [x] This PR is reasonably limited in scope, to help ensure that:
  1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
  2. There isn't too much of a burden on reviewers.
  3. Any problems it causes have a small "blast radius".
  4. It'll be easier to rollback if that becomes necessary.
- [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
- [x] The data dictionary has been updated with any field mapping changes, if any were made.
- [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
- [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
- [x] Reviews are requested from both:
  - At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
  - Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
- [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
  - Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Review by aadedejifearless — APPROVED** [review]
LGTM

**Review by brandoncruz3 — DISMISSED** [review]
🚀

File: ops/terraform/services/base/main.tf [hunk 1]
```diff
@@ -1,18 +1,89 @@
...
+    {
+      Version = "2012-10-17",
+      Id      = "config-key-policy",
+      Statement = [
+        # This policy statement is the default key policy allowing the account principal to delegate
+        # permissions, via IAM policies, to other principals within the AWS Account. Due to this
+        # delegation, it is not necessary to include additional IAM User/Group-specific policy
+        # statements within this key policy, as that permission delegation can be handled via
+        # distinct IAM Policies attached to those entities. See
+        # https://docs.aws.amazon.com/kms/latest/developerguide/key-policy-default.html#key-policy-default-allow-root-enable-iam
```
Comment (mjburling) [inline ops/terraform/services/base/main.tf:26]: Thank you. The key policy/IAM policy divide is a bit of a mess. I'm heartened to see that we're not extending the mess.

File: ops/terraform/services/base/main.tf [hunk 2]
```diff
@@ -1,18 +1,89 @@
...
+          Resource = "*"
+        }
+      ]
+    }
+  )
+  # The KMS key ID/ARN is dependent on whether the current env is ephemeral or not: if it is, then
+  # this Terraservice will use the seed env's config CMK instead of creating one specific to the
+  # environment, otherwise, if the env is established, this Terraservice will encrypt all sensitive
+  # SSM configuration with the primary key defined within this Terraservice
+  kms_key_id = local.is_ephemeral_env ? one(data.aws_kms_key.cmk[*].arn) : one(aws_kms_key.primary[*].arn)
```
Comment (mjburling) [inline ops/terraform/services/base/main.tf:43]: 👍

File: apps/utils/cipher/cipher.main.kts [hunk 3]
```diff
@@ -6,23 +6,13 @@
...
-import com.amazonaws.encryptionsdk.CryptoResult
-import com.amazonaws.encryptionsdk.kmssdkv2.KmsMasterKey
-import com.amazonaws.encryptionsdk.kmssdkv2.KmsMasterKeyProvider
+import com.amazonaws.encryptionsdk.kmssdkv2.AwsKmsMrkAwareMasterKeyProvider
 import kotlinx.collections.immutable.PersistentMap
 import kotlinx.collections.immutable.persistentMapOf
-import okio.Buffer
-import okio.BufferedSource
-import okio.ByteString
+import okio.*
```
Comment (mjburling) [inline apps/utils/cipher/cipher.main.kts:12]: I'd really rather not allow what amounts to a style decision to hold this up any longer, but... maybe it's reversing this isn't so bad if you're using an IDE?

**Review by mjburling — DISMISSED** [review]
Approved with minor reservations.

Overall, this looks exactly as we discussed. I don't see anything obviously wrong and don't have any questions. I really appreciate your including the notes on the steps taken to ensure it's functional. Thanks!

**Review by aadedejifearless — APPROVED** [review]

**Review by mjburling — APPROVED** [review]

**Pull request merged**